### PR TITLE
140034 - BESS CS - Get AVCS Catalogue from SCS

### PR DESCRIPTION
**Pull request by ratnas104142**
This PR consists changes of [PBI 140034](https://dev.azure.com/ukhydro/Exchange%20Set%20Service/_workitems/edit/140034) :

- Code changes for getting AVCS catalogue data from SCS for BESS Configuration Service
- Added unit tests
- Added functional tests

### CODE REVIEW

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Services/ISalesCatalogueService.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
```
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Services/ISalesCatalogueService.cs:4]: remove unused usings.

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs [hunk 2]
```diff
@@ -7,451 +8,582 @@ public enum EventIds
...
+
         /// <summary>
         /// 809109 - Exception occurred while processing bess config frequency
         /// </summary>
         BessConfigFrequencyProcessingException = 809109,
+
+        /// <summary>
+        /// 809110 - Request to Sales Catalogue Service's catalogue essData endpoint is started.
+        /// </summary>
+        SCSGetSalesCatalogueDataRequestStarted = 809110,
```
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs:561]:
```suggestion
        ScsGetSalesCatalogueDataRequestStarted = 809110,
```

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs [hunk 3]
```diff
@@ -7,451 +8,582 @@ public enum EventIds
...
+
+        /// <summary>
+        /// 809111 - Request to Sales Catalogue Service's catalogue essData endpoint is completed.
+        /// </summary>
+        SCSGetSalesCatalogueDataRequestCompleted = 809111,
```
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs:566]:
```suggestion
        ScsGetSalesCatalogueDataRequestCompleted = 809111,
```

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs [hunk 4]
```diff
@@ -7,451 +8,582 @@ public enum EventIds
...
+
         /// <summary>
-        /// 809110 - Bess Builder Service Started
+        /// 809112 - Request to Sales Catalogue Service's catalogue essData endpoint failed due to Non-Ok response.
         /// </summary>
-        BessBuilderServiceStarted = 809110,
+        SalesCatalogueServiceCatalogueDataNonOkResponse = 809112,
```
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs:571]:
```suggestion
        ScsGetSalesCatalogueDataNonOkResponse = 809112,
```

**Review by Nadeem14456 — CHANGES_REQUESTED** [review]

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs [hunk 5]
```diff
@@ -7,451 +8,582 @@ public enum EventIds
...
+
+        /// <summary>
+        /// 809113 - Request for retrying sales catalogue service endpoint.
+        /// </summary>
+        RetryHttpClientSCSRequest = 809113,
```
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs:576]:
```suggestion
        RetryHttpClientScsRequest = 809113,
```

**Review by harshal11869 — CHANGES_REQUESTED** [review]

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Models/Scs/Response/SalesCatalogueDataResponse.cs [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+{
+    public class SalesCatalogueDataResponse
+    {
+        public List<SalesCatalogueDataProductResponse> ResponseBody { get; set; }
+        public HttpStatusCode ResponseCode { get; set; }
+        public DateTime? LastModified { get; set; }
+        public object Clone()
+        {
+            return this.MemberwiseClone();
+        }
```
[the remark this answers is not in the document]
Comment (ratnas104142) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Models/Scs/Response/SalesCatalogueDataResponse.cs]: Resolved!

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/SalesCatalogueClient.cs [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+{
+    [ExcludeFromCodeCoverage]
+    public class SalesCatalogueClient : ISalesCatalogueClient
+    {
+        private readonly HttpClient httpClient;
+
+        public SalesCatalogueClient(HttpClient httpClient)
+        {
+            this.httpClient = httpClient;
+        }
```
[the remark this answers is not in the document]
Comment (ratnas104142) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/SalesCatalogueClient.cs]: Resolved!

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/SalesCatalogueClient.cs [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+            if (requestBody != null)
+                content = new StringContent(requestBody, Encoding.UTF8, "application/json");
+
+            using var httpRequestMessage = new HttpRequestMessage(method, uri)
+            { Content = content };
+
+            httpRequestMessage.Headers.Add("X-Correlation-ID", CommonHelper.CorrelationID.ToString());
+
+            httpRequestMessage.Headers.Authorization = new AuthenticationHeaderValue("Bearer", authToken);
```
[the remark this answers is not in the document]
Comment (ratnas104142) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/SalesCatalogueClient.cs]: Resolved!

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs (hunk 4 above)
Comment (ratnas104142) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs]: Resolved!

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs (hunk 3 above)
Comment (ratnas104142) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs]: Resolved!

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs (hunk 2 above)
Comment (ratnas104142) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs]: Resolved!

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Services/ISalesCatalogueService.cs (hunk 1 above)
Comment (ratnas104142) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Services/ISalesCatalogueService.cs]: Resolved!

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Services/SalesCatalogueServiceTests.cs [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+            salesCatalogueService = new SalesCatalogueService(fakeLogger, fakeSaleCatalogueConfig, fakeAuthScsTokenProvider, fakeSalesCatalogueClient);
+        }
+
+        #region GetSalesCatalogueDataProductResponse
+        private List<SalesCatalogueDataProductResponse> GetSalesCatalogueDataProductResponse()
+        {
+            return
+                new List<SalesCatalogueDataProductResponse>()
+                {
+                    new SalesCatalogueDataProductResponse()
```
[the remark this answers is not in the document]
Comment (ratnas104142) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Services/SalesCatalogueServiceTests.cs]: Resolved!

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Services/ConfigurationServiceTests.cs [hunk 10]
```diff
@@ -381,5 +392,28 @@ private List<BessConfig> GetFakeConfigurationSettingNotEnabled()
...
+
+        #region GetSalesCatalogueDataProductResponse
+        private SalesCatalogueDataResponse GetSalesCatalogueDataResponse()
+        {
+            return new SalesCatalogueDataResponse
+            {
+                ResponseCode = HttpStatusCode.OK,
+                ResponseBody = new List<SalesCatalogueDataProductResponse>()
+                {
+                    new SalesCatalogueDataProductResponse
```
[the remark this answers is not in the document]
Comment (ratnas104142) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Services/ConfigurationServiceTests.cs]: Resolved!

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/appsettings.json [hunk 11]
```diff
@@ -32,5 +32,17 @@
     "BessStorageConfiguration": {
         "ConnectionString": "",
         "ContainerName": "bess-configs"
+    },
+    "SalesCatalogue": {
```
[the remark this answers is not in the document]
Comment (ratnas104142) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/appsettings.json]: Resolved!

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/appsettings.local.overrides.json [hunk 12]
```diff
@@ -32,5 +32,17 @@
     "BessStorageConfiguration": {
         "ConnectionString": "",
         "ContainerName": "bess-configs"
+    },
+    "SalesCatalogue": {
```
[the remark this answers is not in the document]
Comment (ratnas104142) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/appsettings.local.overrides.json]: Resolved!

File: UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Helpers/AuthTokenProvider.cs [hunk 13]
```diff
@@ -0,0 +1,59 @@
+﻿using Microsoft.Identity.Client;
+
+namespace UKHO.BESS.API.FunctionalTests.Helpers
+{
+    public class AuthTokenProvider
+    {
+        static string? ScsAccessToken;
```
Comment (harshal11869) [inline UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Helpers/AuthTokenProvider.cs:7]: Any reason to make string nullable? As string accepts null as well.

**Review by harshal11869 — CHANGES_REQUESTED** [review]

**Review by Nadeem14456 — APPROVED** [review]
LGTM.

**Review by harshal11869 — APPROVED** [review]
LGTM.

**Review by Mayuresh-Gawde — APPROVED** [review]
LGTM from QA!!

**Review by DanOwenUKHO — APPROVED** [review]

**Pull request merged**